File: app/src/main/java/com/example/interactions/AbstractWebViewFragment.java
package com.example.interactions;

import android.annotation.SuppressLint;
import android.databinding.DataBindingUtil;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.example.R;
import com.example.databinding.WebviewBinding;

public abstract class AbstractWebViewFragment extends AbstractFragment {
	private boolean isSet = false;
	protected WebviewBinding binding;

	private WebViewClient webviewClient = new WebViewClient() {
		private boolean error;

		@Override
		public void onPageStarted(WebView webview, String url, Bitmap favicon) {
			error = false;
		}

		@Override
		public void onPageFinished(WebView webview, String url) {
			if (!error) {
				onPageLoaded(webview, url);
			}
		}

		@Override
		public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
			super.onReceivedError(view, request, error);
			this.error = true;
		}

		@Override
		public boolean shouldOverrideUrlLoading(WebView webview, String url) {
			return (shouldOverride(webview, url) && onHandleCallback(webview, url))
				|| super.shouldOverrideUrlLoading(webview, url);
		}

		@Override
		public void onFormResubmission(WebView view, Message dontResend, Message resend) {
			//resend POST request without confirmation.
			resend.sendToTarget();
		}
	};

	@Nullable
	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		super.onCreateView(inflater, container, savedInstanceState);

		binding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.webview, container, false);

		setupWebView();
		navigateToInitialPage(binding.webview);

		return binding.getRoot();
	}

	@SuppressLint("SetJavaScriptEnabled")
	private void setupWebView() {
		if (isSet) {
			return;
		}

		WebSettings settings = binding.webview.getSettings();

		if (settings != null) {
			settings.setJavaScriptEnabled(true);
		}

		binding.webview.setHorizontalScrollBarEnabled(false);
		binding.webview.setWebViewClient(webviewClient);

		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
			WebView.setWebContentsDebuggingEnabled(true);
		}
		if (Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN) {
			//refs: https://code.google.com/p/android/issues/detail?id=35288
			binding.webview.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
		}

		isSet = true;
	}

	protected WebView getWebview() {
		setupWebView();
		return binding.webview;
	}

	protected abstract void navigateToInitialPage(WebView webview);

	protected void onPageLoaded(WebView webview, String url) {
	}

	protected boolean shouldOverride(WebView webview, String url) {
		return false;
	}

	protected boolean onHandleCallback(WebView webview, String url) {
		return false;
	}
}

File: libs/toolkit-android/src/main/kotlin/br/com/vitorsalgado/example/utils/base/AbstractWebViewFragment.kt
// package br.com.vitorsalgado.example.features
//
// import android.annotation.SuppressLint
// import android.annotation.TargetApi
// import android.databinding.DataBindingUtil
// import android.graphics.Bitmap
// import android.os.Build
// import android.os.Bundle
// import android.os.Message
// import android.view.LayoutInflater
// import android.view.View
// import android.view.ViewGroup
// import android.webkit.WebResourceError
// import android.webkit.WebResourceRequest
// import android.webkit.WebView
// import android.webkit.WebViewClient
// import br.com.vitorsalgado.example.R
// import br.com.vitorsalgado.example.databinding.WebviewBinding
//
// abstract class AbstractWebViewFragment : AbstractFragment() {
//  private var isSet = false
//  private val webviewClient = object : WebViewClient() {
//    private var error: Boolean = false
//
//    override fun onPageStarted(webview: WebView, url: String, favicon: Bitmap) {
//      error = false
//    }
//
//    override fun onPageFinished(webview: WebView, url: String) {
//      if (!error) {
//        onPageLoaded(webview, url)
//      }
//    }
//
//    override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
//      super.onReceivedError(view, request, error)
//      this.error = true
//    }
//
//    override fun shouldOverrideUrlLoading(webview: WebView, url: String): Boolean {
//      return shouldOverride(webview, url) && onHandleCallback(webview, url) || super.shouldOverrideUrlLoading(webview, url)
//    }
//
//    @TargetApi(Build.VERSION_CODES.N)
//    override fun shouldOverrideUrlLoading(webview: WebView, request: WebResourceRequest): Boolean {
//      val uri = request.url.toString()
//
//      return shouldOverride(webview, uri) && onHandleCallback(webview, uri) || super.shouldOverrideUrlLoading(webview, uri)
//    }
//
//
//    override fun onFormResubmission(view: WebView, dontResend: Message, resend: Message) {
//      //resend POST request without confirmation.
//      resend.sendToTarget()
//    }
//  }
//
//  protected val webview: WebView
//    get() {
//      setupWebView()
//      return webview
//    }
//
//  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
//    super.onCreateView(inflater, container, savedInstanceState)
//
//    binding = DataBindingUtil.inflate(layoutInflater, R.layout.webview, container, false)
//
//    setupWebView()
//    navigateToInitialPage(binding.webview)
//
//    return binding.root
//  }
//
//  @SuppressLint("SetJavaScriptEnabled")
//  private fun setupWebView() {
//    if (isSet) {
//      return
//    }
//
//    val settings = binding.webview.settings
//
//    if (settings != null) {
//      settings.javaScriptEnabled = true
//    }
//
//    binding.webview.isHorizontalScrollBarEnabled = false
//    binding.webview.webViewClient = webviewClient
//
//    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
//      WebView.setWebContentsDebuggingEnabled(true)
//    }
//    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN) {
//      //refs: https://code.google.com/p/android/issues/detail?id=35288
//      binding.webview.setLayerType(View.LAYER_TYPE_SOFTWARE, null)
//    }
//
//    isSet = true
//  }
//
//  protected abstract fun navigateToInitialPage(webview: WebView)
//
//  protected fun onPageLoaded(webview: WebView, url: String) {}
//
//  protected fun shouldOverride(webview: WebView, url: String): Boolean {
//    return false
//  }
//
//  protected fun onHandleCallback(webview: WebView, url: String): Boolean {
//    return false
//  }
// }
